src/main/java/org/prebid/server/bidder/ucfunnel/UcfunnelBidder.java

```diff
@@ -0,0 +1,161 @@
+        final List<BidderError> errors = new ArrayList<>();
+
+        if (CollectionUtils.sizeIsEmpty(request.getImp())) {
+            errors.add(BidderError.badInput("No valid impressions in the bid request"));
+            return Result.of(Collections.emptyList(), errors);
+        }
+
+        String partnerId = null;
+        try {
+            final ExtImpUcfunnel extImpUcfunnel = parseImpExt(request.getImp().get(0));
+            final String adUnitId = extImpUcfunnel.getAdunitid();
+            partnerId = extImpUcfunnel.getPartnerid();
+            if (StringUtils.isEmpty(partnerId) || StringUtils.isEmpty(adUnitId)) {
+                errors.add(BidderError.badInput("No PartnerId or AdUnitId in the bid request"));
+                return Result.of(Collections.emptyList(), errors);
+            }
+        } catch (PreBidException e) {
+            errors.add(BidderError.badInput(e.getMessage()));
+        }
+
+        final String body = mapper.encode(request);
+        final String requestUrl = endpointUrl + "/" + HttpUtil.encodeUrl(partnerId) + "/request";
+
+        return Result.of(Collections.singletonList(
+                HttpRequest.<BidRequest>builder()
+                        .method(HttpMethod.POST)
+                        .uri(requestUrl)
+                        .body(body)
+                        .headers(HttpUtil.headers())
+                        .payload(request)
+                        .build()),
+                errors);
+    }
+
+    private ExtImpUcfunnel parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), UCFUNNEL_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
+        final int statusCode = httpCall.getResponse().getStatusCode();
+        if (statusCode == HttpResponseStatus.NO_CONTENT.code()) {
+            return Result.of(Collections.emptyList(), Collections.emptyList());
+        } else if (statusCode == HttpResponseStatus.BAD_REQUEST.code()) {
+            return Result.emptyWithError(BidderError.badInput("bad request"));
+        } else if (statusCode != HttpResponseStatus.OK.code()) {
+            return Result.emptyWithError(BidderError.badServerResponse(String.format("Unexpected HTTP status %s.",
+                    statusCode)));
+        }
+
+        try {
+            final BidResponse bidResponse = decodeBodyToBidResponse(httpCall);
+            final List<BidderBid> bidderBids = bidResponse.getSeatbid().stream()
+                    .filter(Objects::nonNull)
+                    .map(SeatBid::getBid)
+                    .filter(Objects::nonNull)
+                    .flatMap(Collection::stream)
+                    .map(bid -> BidderBid.of(bid, validateBidType(bid.getImpid(), bidRequest.getImp()),
```

Comment:
bc of validateBidType (btw I'd prefer validat**ed**BidType) you can receive BidderBid with empty bidTupe (null)